### PR TITLE
Document ledger invariants

### PR DESCRIPTION
**Pull request by jumaffre**
These invariants were not documented anywhere, even though we try very hard/have tests to verify those, mostly for auditor and operator convenience.

### CODE REVIEW

File: doc/operations/ledger_snapshot.rst [hunk 1]
```diff
@@ -86,3 +86,15 @@ A low value for ``ledger.chunk_size`` means that smaller ledger files are genera
 Similarly, a low value for ``snapshots.tx_count`` means that snapshots are generated often and that join/recovery time will be short, at the cost of additional workload on the primary node for snapshot generation.
 
 .. tip:: Uncommitted ledger files (which are likely to contain committed transactions) can also be used on join/recovery, as long as they are copied to the node's ``ledger.directory`` directory.
+
+Invariants
+----------
+
+1. To facilitate audit and verification of the integrity of the ledger, individual ledger files always end on a signature transaction.
+
+2. For operator convenience, all committed ledger files (``.committed`` suffix) are the same on all up-to-date nodes.
```
Comment (eddyashton) [inline doc/operations/ledger_snapshot.rst:95]: I think this could be slightly clearer/stronger. ie all committed transactions are present in exactly 1 `.committed` file (no duplicates), `.committed` files start and end at the same seqnos on every node (consistent chunking), and files with the same name on different nodes are _identical_ (no node-specific headers/contents in ledger). All caveated by "on the nodes where these files are present", or "all up-to-date nodes", or some other way of describing this.

Comment (jumaffre) [reply doc/operations/ledger_snapshot.rst]: Good points. Wrt "all committed transactions are present in exactly 1 .committed file", I think this is bit confusing as 
since some committed transactions will be stored in non-committed chunks (explained at the bottom of this section (https://microsoft.github.io/CCF/main/operations/ledger_snapshot.html#file-layout). I've added the other points as sub-points in the `2.` invariant.

File: doc/operations/ledger_snapshot.rst [hunk 2]
```diff
@@ -86,3 +86,18 @@ A low value for ``ledger.chunk_size`` means that smaller ledger files are genera
...
+----------
+
+1. To facilitate audit and verification of the integrity of the ledger, individual ledger files always end on a signature transaction.
+
+2. For operator convenience, all committed ledger files (``.committed`` suffix) are the same on all up-to-date nodes. More precisely, among up-to-date nodes:
+
+- Committed ledger files start and end at the same ``seqno``.
+- Committed ledger files with the same name are byte-identical.
+
+3. Snapshots are always taken at the ``seqno`` of a signature transaction.
```
Comment (wintersteiger) [inline doc/operations/ledger_snapshot.rst:100]: Maybe rephrase to "snapshots end at a signature", to make it clear that not every signature produces a snapshot?

File: doc/operations/ledger_snapshot.rst [hunk 3]
```diff
@@ -86,3 +86,18 @@ A low value for ``ledger.chunk_size`` means that smaller ledger files are genera
 Similarly, a low value for ``snapshots.tx_count`` means that snapshots are generated often and that join/recovery time will be short, at the cost of additional workload on the primary node for snapshot generation.
 
 .. tip:: Uncommitted ledger files (which are likely to contain committed transactions) can also be used on join/recovery, as long as they are copied to the node's ``ledger.directory`` directory.
```
Comment (wintersteiger) [inline doc/operations/ledger_snapshot.rst:88]: Should we start adjusting that to exclude temporary recovery files?

**Review by wintersteiger — APPROVED** [review]

File: doc/operations/ledger_snapshot.rst (hunk 3 above)
Comment (jumaffre) [reply doc/operations/ledger_snapshot.rst:88]: Yes, as soon as the feature is on `main`, we should reflect this here.
